Stabilise and improve keynote/numbers/pages attachment tests

squelart opened the pull request:
#### 0c97b9e472f42b3ed1de7293edaf792b7b3bba07
<pre>
Stabilise and improve keynote/numbers/pages attachment tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=262987">https://bugs.webkit.org/show_bug.cgi?id=262987</a>
rdar://116777261

Reviewed by Aditya Keerthi.

Attachment-{keynote,numbers,pages}.html were flaky, with the actual and/or
expected rendering sometimes missing the icon, because that icon is only
fetched during the first rendering.

* LayoutTests/fast/attachment/mac/attachment-keynote.html:
* LayoutTests/fast/attachment/mac/attachment-numbers.html:
* LayoutTests/fast/attachment/mac/attachment-pages.html:
Use reftest-wait and takeScreenshotWhenAttachmentsSettled to delay the
test screenshot until there is an icon.

* LayoutTests/fast/attachment/mac/attachment-keynote-expected-mismatch.html:
* LayoutTests/fast/attachment/mac/attachment-numbers-expected-mismatch.html:
* LayoutTests/fast/attachment/mac/attachment-pages-expected-mismatch.html:
Instead of comparing a mismatch against nothing at all, compare with
an attachment with just a different icon.

* LayoutTests/fast/attachment/mac/attachment-keynote-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-keynote-icon.html: Added.
* LayoutTests/fast/attachment/mac/attachment-numbers-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-numbers-icon.html: Added.
* LayoutTests/fast/attachment/mac/attachment-pages-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-pages-icon.html: Added.
New tests that check that the keynote/numbers/pages icons can be set
through the title or the file attribute.

* LayoutTests/fast/attachment/resources/test-file.key: Added.
* LayoutTests/fast/attachment/resources/test-file.numbers: Added.
* LayoutTests/fast/attachment/resources/test-file.pages: Added.
Empty files that can be referenced from the above tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
Removed gardening changes.

Canonical link: <a href="https://commits.webkit.org/269840@main">https://commits.webkit.org/269840@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf34b19c05b25d513cc1dd2bf293f39d942ce5a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1845 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21900 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3436 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22438 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1394 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26480 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1174 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21436 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27692 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21647 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21711 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25461 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1139 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18835 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1129 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21209 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5678 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1551 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1441 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/attachment/mac/attachment-keynote-expected.html:
```diff
@@ -0,0 +1,18 @@
+
+var attachments = document.getElementsByTagName("attachment");
+for (var i = 0; i < attachments.length; i++)
+    attachments[i].file = file;
```
Looking at these tests, the empty Keynote/Pages/Numbers, and the fact that the previous files were `-expected-mismatch.html`, my understanding is that this previously tested the fact that attachment elements drew something on the page.

Now, these tests appear to be testing that setting the file programmatically, and with an attribute, have the same result. Is that understanding correct? If so, I think we should preserve the old test behaviors while also adding new tests.

nt1m replied:
I agree with Aditya, it's probably more useful to have an -expected-mismatch.html file that is the exact same attachment but with a different icon.

squelart replied:
Thanks for the reviews.
Yeah, I think I got carried away by previous patches (that were testing e.g. `.txt` icon set two ways), and I didn't quite get the importance of checking that we were drawing *any*thing. 😅
Alright, I'll see if I can make the old mismatch tests stable (and compare with another type of attachment, good idea @nt1m), and I'll add these new ones as well.

nt1m commented on LayoutTests/fast/attachment/mac/attachment-keynote-icon-expected.html:
```diff
@@ -0,0 +1,18 @@
+
+var attachments = document.getElementsByTagName("attachment");
+for (var i = 0; i < attachments.length; i++)
+    attachments[i].file = file;
```
we probably don't need to loop over a single item here?

`document.querySelector("attachment").file = file;`

squelart replied:
That's from the original file. Maybe it used to test multiple attachments? Anyway, I guess I can simplify all of them since I'm heavily modifying the expected files and creating copies.

squelart replied:
Spoke too fast, it's actually from other similar tests I pilfered, and some of them do have multiple attachments. I'll focus on iWork tests in this patch.

nt1m commented on LayoutTests/fast/attachment/mac/attachment-pages-icon-expected.html:
```diff
@@ -0,0 +1,18 @@
+
+var attachments = document.getElementsByTagName("attachment");
+for (var i = 0; i < attachments.length; i++)
+    attachments[i].file = file;
```
ditto

pxlcoder reviewed (approved):
Looks good, with Tim's comments addressed!

squelart committed:
Stabilise and improve keynote/numbers/pages attachment tests
https://bugs.webkit.org/show_bug.cgi?id=262987
rdar://116777261

Reviewed by Aditya Keerthi.

Attachment-{keynote,numbers,pages}.html were flaky, with the actual and/or
expected rendering sometimes missing the icon, because that icon is only
fetched during the first rendering.

* LayoutTests/fast/attachment/mac/attachment-keynote.html:
* LayoutTests/fast/attachment/mac/attachment-numbers.html:
* LayoutTests/fast/attachment/mac/attachment-pages.html:
Use reftest-wait and takeScreenshotWhenAttachmentsSettled to delay the
test screenshot until there is an icon.

* LayoutTests/fast/attachment/mac/attachment-keynote-expected-mismatch.html:
* LayoutTests/fast/attachment/mac/attachment-numbers-expected-mismatch.html:
* LayoutTests/fast/attachment/mac/attachment-pages-expected-mismatch.html:
Instead of comparing a mismatch against nothing at all, compare with
an attachment with just a different icon.

* LayoutTests/fast/attachment/mac/attachment-keynote-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-keynote-icon.html: Added.
* LayoutTests/fast/attachment/mac/attachment-numbers-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-numbers-icon.html: Added.
* LayoutTests/fast/attachment/mac/attachment-pages-icon-expected.html: Added.
* LayoutTests/fast/attachment/mac/attachment-pages-icon.html: Added.
New tests that check that the keynote/numbers/pages icons can be set
through the title or the file attribute.

* LayoutTests/fast/attachment/resources/test-file.key: Added.
* LayoutTests/fast/attachment/resources/test-file.numbers: Added.
* LayoutTests/fast/attachment/resources/test-file.pages: Added.
Empty files that can be referenced from the above tests.

* LayoutTests/platform/mac-wk2/TestExpectations:
Removed gardening changes.

Canonical link: https://commits.webkit.org/269840@main
